llpc/lower/llpcSpirvLowerMath.cpp

```diff
@@ -80,14 +80,35 @@ void SpirvLowerMath::init(Module &module) {
   if (m_shaderStage == ShaderStageInvalid)
     return;
 
-  auto commonShaderMode = Pipeline::getCommonShaderMode(module, getLgcShaderStage(m_shaderStage));
-  m_fp16DenormFlush = commonShaderMode.fp16DenormMode == FpDenormMode::FlushOut ||
-                      commonShaderMode.fp16DenormMode == FpDenormMode::FlushInOut;
-  m_fp32DenormFlush = commonShaderMode.fp32DenormMode == FpDenormMode::FlushOut ||
-                      commonShaderMode.fp32DenormMode == FpDenormMode::FlushInOut;
-  m_fp64DenormFlush = commonShaderMode.fp64DenormMode == FpDenormMode::FlushOut ||
-                      commonShaderMode.fp64DenormMode == FpDenormMode::FlushInOut;
-  m_fp16RoundToZero = commonShaderMode.fp16RoundMode == FpRoundMode::Zero;
+  // NOTE: We try to set denormal-fp-math here because later optimization passes will detect the attributes and decide
+  // what to do. Such attributes will be set once again in LGC.
+  auto shaderMode = Pipeline::getCommonShaderMode(module, getLgcShaderStage(m_shaderStage));
+  m_entryPoint->addFnAttr("denormal-fp-math-f32", "preserve-sign");
+
+  if (shaderMode.fp16DenormMode == FpDenormMode::FlushNone || shaderMode.fp16DenormMode == FpDenormMode::FlushIn ||
+      shaderMode.fp64DenormMode == FpDenormMode::FlushNone || shaderMode.fp64DenormMode == FpDenormMode::FlushIn) {
+    m_entryPoint->addFnAttr("denormal-fp-math", "ieee");
+  } else if (shaderMode.fp16DenormMode == FpDenormMode::FlushOut ||
+             shaderMode.fp16DenormMode == FpDenormMode::FlushInOut ||
+             shaderMode.fp64DenormMode == FpDenormMode::FlushOut ||
+             shaderMode.fp64DenormMode == FpDenormMode::FlushInOut) {
+    m_entryPoint->addFnAttr("denormal-fp-math", "preserve-sign");
+  }
+
+  if (shaderMode.fp32DenormMode == FpDenormMode::FlushNone || shaderMode.fp32DenormMode == FpDenormMode::FlushIn) {
+    m_entryPoint->addFnAttr("denormal-fp-math-f32", "ieee");
+  } else if (shaderMode.fp32DenormMode == FpDenormMode::FlushOut ||
+             shaderMode.fp32DenormMode == FpDenormMode::FlushInOut) {
+    m_entryPoint->addFnAttr("denormal-fp-math-f32", "preserve-sign");
```

Comment:
Without going through all the cases, isn't this attribute always set because of LOC 86?

Reply:
This is copied from PatchSetupTargetFeatures::setupTargetFeatures. I think we can remove this branch since "denormal-fp-math-f32"="preserve-sign" will be set unconditionally as initial attributes.